### PR TITLE
Handle coins and proposals

### PR DESCRIPTION
**Pull request by nodar-chkuaselidze**
Proposals will keep track of the coins updates which happen on `tx`, `confirmed`, `unconfirmed`, `remove tx` events and will reject proposals (and free locked coins) accordingly.

* Proposal: add equals method.
* WalletClient: listen for events
* WalletNullClient - mostly for testing
* Watch for walletdb events
* Save Outpoint -> PID mapping in DB.
* AddTX - Handle double spend in mempool or unseen transaction in block.
* confirmedTX - Handle double spend if we got here somehow.
* removeTX - Handle reorg double spend or mempool double spend of coins
* coin spend - Addtx/confirmed will also handle case when transaction was broadcasted from different node (Basically same as double spend)
* Test double spend scenarios.
* Cleanup some tests.

### CODE REVIEW

File: lib/proposal.js [hunk 1]
```diff
@@ -13,27 +13,39 @@ const {Outpoint, TX} = bcoin;
...
-  'REORG',
+  'DBLSPEND',
   'VERIFY'
 ];
 
 const statusMessages = [
   'Proposal is in progress.',
+
   'Proposal has been approved.',
+
```
Comment (bucko13) [inline lib/proposal.js]: Unnecessary whitespace?

File: lib/proposaldb.js [hunk 2]
```diff
@@ -265,13 +265,29 @@ class ProposalDB {
...
+   * @param {Outpoint} outpoint
+   * @returns {Promise<Number>} Proposal ID
+   */
+
+  getPIDByOutpoint(outpoint) {
+    return Proposal.getPIDByOutpoint(this.bucket, outpoint);
+  }
+
   /**
    * Get proposal coins (use only on pending proposals)
```
Comment (bucko13) [inline lib/proposaldb.js]: There is a status check for proposals right? Should this start with a check to confirm it's pending?

File: lib/proposaldb.js [hunk 3]
```diff
@@ -485,6 +508,123 @@ class ProposalDB {
...
+   * Transaction was added in mempool or chain.
+   * Check if we have proposal using same Coins.
+   * reject if necessary.
+   * @private
+   * @param {bcoin#TX} tx
+   * @param {bcoin#TXDB#Details} details
+   */
+
+  addTX(tx, details) {
+    return this.rejectProposalByTX(tx);
```
Comment (bucko13) [inline lib/proposaldb.js]: What is the purpose of this (and the next) method? `details` isn't used, and this doesn't really add a TX to anything, but just seems to run `rejectProposalByTX`. Is this a placeholder for later API development?

File: lib/proposal.js (hunk 1 above)
Comment (nodar-chkuaselidze) [reply lib/proposal.js]: Separated rejections.

File: lib/proposaldb.js (hunk 3 above)
Comment (nodar-chkuaselidze) [reply lib/proposaldb.js]: This is called by event `tx` which means transaction was added to either mempool or chain. I thought comment for method was good enough.  :)

File: lib/proposal.js (hunk 1 above)
Comment (nodar-chkuaselidze) [reply lib/proposal.js]: separated rejections.

File: lib/proposaldb.js (hunk 3 above)
Comment (nodar-chkuaselidze) [reply lib/proposaldb.js]: This is called by event tx which means transaction was added to either mempool or chain. I thought comment for method was good enough. :)

File: lib/proposaldb.js (hunk 2 above)
Comment (nodar-chkuaselidze) [reply lib/proposaldb.js]: I might mark this method private, this is only used in `approveProposal` method which already checks whether PID is approved.
